rust/compaction/src/store.rs

```diff
@@ -0,0 +1,886 @@
+//! This module contains a wrapper for an [`ObjectStore`] that adds logging functionality and customised multipart upload
+//! buffering size.
+//!
+use bytes::{Bytes, BytesMut};
+/*
+ * Copyright 2022-2024 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+use futures::{stream::BoxStream, Future};
+use log::info;
+use num_format::{Locale, ToFormattedString};
+use object_store::{
+    path::Path, GetOptions, GetRange, GetResult, ListResult, MultipartUpload, ObjectMeta,
+    ObjectStore, PutMultipartOpts, PutOptions, PutPayload, PutResult, Result, UploadPart,
+};
+use std::{
+    pin::Pin,
+    sync::{Arc, Mutex},
+};
+
+pub const MULTIPART_BUF_SIZE: usize = 20 * 1024 * 1024;
+
+/// Simple struct for storing various statistics about the operation of the store.
+#[derive(Debug, Eq, PartialOrd, Ord, PartialEq, Clone)]
+struct LoggingData {
+    /// The number of GET requests logged.
+    get_count: usize,
+    /// The total number of bytes read across all files.
+    get_bytes_read: usize,
+    /// Multipart upload buffer capacity
+    capacity: usize,
+}
+
+impl Default for LoggingData {
+    fn default() -> Self {
+        Self {
+            get_count: Default::default(),
+            get_bytes_read: Default::default(),
+            capacity: MULTIPART_BUF_SIZE,
+        }
+    }
+}
+
+/// Allow setting a size hint for multipart uploads.
+pub trait MultipartSizeHintable: ObjectStore {
```

Comment:
I found this name quite confusing when I was reading the S3 code, as it's not clear it's an object store. It's a bit long but it might be better as MultipartSizeHintableObjectStore?